internal/controller/driver_controller.go

```diff
@@ -18,45 +18,1033 @@ package controller
+	RbdDriverType,
+	CephFsDriverType,
+	NfsDriverType,
+))
+
 // DriverReconciler reconciles a Driver object
 type DriverReconciler struct {
 	client.Client
 	Scheme *runtime.Scheme
 }
 
-//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers,verbs=get;list;watch;create;update;patch;delete
-//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers/status,verbs=get;update;patch
-//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers/finalizers,verbs=update
+// A local reconcile object tied to a single reconcile iteration
+type driverReconcile struct {
+	DriverReconciler
+
+	ctx          context.Context
+	log          logr.Logger
+	driver       csiv1a1.Driver
+	driverPrefix string
+	driverType   DriverType
+	images       map[string]string
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *DriverReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	// Define conditions for an OperatorConfig change that the require queuing of reconciliation
+	// request for drivers
+	driverDefaultsPredicate := builder.WithPredicates(
+		predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool {
+				opConf, ok := e.Object.(*csiv1a1.OperatorConfig)
+				return ok && opConf.Spec.DriverSpecDefaults != nil
+			},
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				old, oldOk := e.ObjectOld.(*csiv1a1.OperatorConfig)
+				new, newOk := e.ObjectNew.(*csiv1a1.OperatorConfig)
+				return oldOk && newOk &&
+					!reflect.DeepEqual(old.Spec.DriverSpecDefaults, new.Spec.DriverSpecDefaults)
+			},
+			DeleteFunc: func(e event.DeleteEvent) bool {
+				opConf, ok := e.Object.(*csiv1a1.OperatorConfig)
+				return ok && opConf.Spec.DriverSpecDefaults != nil
+			},
+			GenericFunc: func(event.GenericEvent) bool {
+				return false
+			},
+		},
+	)
+
+	// Enqueue a reconcile request for all existing drivers, used to trigger a reconcile
+	// for all drivers  whenever the driver default configuration changes
+	enqueueAllDrivers := handler.EnqueueRequestsFromMapFunc(
+		func(_ context.Context, obj client.Object) []reconcile.Request {
+			driverList := csiv1a1.DriverList{}
+			if err := r.List(context.Background(), &driverList); err != nil {
+				return []reconcile.Request{}
+			}
+
+			requests := make([]reconcile.Request, len(driverList.Items))
+			for i := range driverList.Items {
+				requests[i].NamespacedName = client.ObjectKeyFromObject(&driverList.Items[i])
+			}
+			return requests
+		},
+	)
+
+	// Enqueue a reconcile request based on an annotation marking a soft ownership
+	enqueueFromOwnerRefAnnotation := handler.EnqueueRequestsFromMapFunc(
+		func(_ context.Context, obj client.Object) []reconcile.Request {
+			ownerRef := obj.GetAnnotations()[ownerRefAnnotationKey]
+			if ownerRef == "" {
+				return nil
+			}
+
+			ownerObjKey := client.ObjectKey{}
+			if err := json.Unmarshal([]byte(ownerRef), &ownerObjKey); err != nil {
+				return nil
+			}
+
+			return []reconcile.Request{{
+				NamespacedName: ownerObjKey,
+			}}
```

Comment:
Not sure I follow, how many events do you expect to see on a CSIDriver resource? Also compared to a k8s API cal marshaling/unmarshaling such a small struct is not an heavy operation

Reply:
I should've reversed my statement, my intention is for tracking a namespace we don't need json for tracking it right?

Reply:
Technically we can write code that only has a namespace as of the assumption that the name will be identical between the Ceph CSI driver and the k8s CSI driver. But please consider that this assumption might change in the future. Let me give you an example:

Let's say we get a requirement to allow the renaming of the CSIDriver CR. And assume there is a real reason why we cannot just uninstall the original driver and install a new one. In that case, I would imagine we will add a spec field to capture the name of the k8s CSI driver we want to create. But that will lead back to a need for both name and namespace.  

If we do not take that into consideration today we will get into a situation where we will need to either live with code that handles both options or code that needs to "upgrade" existing owner refs . Both are bad options and I am willing to take the almost non-existent performance hit in order to future-proof against these kinds of requirements.

Reply:
The example presented here as a lot of indirect assumptions which is hard to implement. I'm not talking about performance. However, this isn't holding up anything, could be resolved.

Reply:
@leelavg Wil resolve.
But can you please point out the indirect assumptions?

Reply:
> we get a requirement to allow the renaming of the CSIDriver CR
- this requirement seems to me as an assumption which is hard to implement if at all possible.

Reply:
It is not an assumption if you look at the surrounding text, it is brought up as a potential enhancement in case such a requirement will come up. 

And it is not impossible, it just requires us to decouple the `driver.csi.ceph.io` name from the `csidriver` name. This can easily be accomplished via a spec field that defaults to the CR name if not specified.